File: 10.5_My_Photos_Asyncronous/app/src/main/java/com/example/a55myphotos/network/RetrofitClient.java
package com.example.a55myphotos.network;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitClient {

    private static Retrofit retrofit;
    private static final String BASE_URL = "https://jsonplaceholder.typicode.com";

    public static Retrofit getRetrofit() {
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        return retrofit;
    }

    /*private RetrofitClient INSTANCE;
    // Have only 1 instance
    private Retrofit retrofit;
    private static final String BASE_URL = "https://jsonplaceholder.typicode.com/";

    // Lazy Initialization
    public RetrofitClient getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new RetrofitClient();
            // Builder design pattern

        }
        return INSTANCE;
    }

    public Retrofit getRetrofit() {
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .build();
        }
        return retrofit;
    }

    private RetrofitClient() {
    }*/
}

File: 9.6_WeatherApplication/app/src/main/java/com/example/a96_weatherapplication/network/RetrofitClient.kt
package com.example.a96_weatherapplication.network

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitClient {
    val BASE_URL = "https://api.weatherbit.io/v2.0/forecast/"
    var instance: Retrofit? = null
        /// any of this  will work
        get() {
            if (field == null) {
                field = Retrofit.Builder().baseUrl(BASE_URL)
                    .addConverterFactory(
                        GsonConverterFactory
                            .create()
                    ).build()
            }
            return field
        }


}